### ISSUE TITLE
Reconsider `@context` injection feature based on implementer feedback

### ISSUE THREAD

**Issue by msporny** [w3c/vc-data-integrity#281]
After performing an implementation, and working through a PR, @rflechtner has proposed that we remove `@context` injection feature from the Data Integrity specification in this comment:

https://github.com/w3c/vc-data-integrity/issues/231#issuecomment-2209299341

The feature was added based on feedback that "implementers should be able to inject a context". The current text doesn't quite do that; it's a bit of a half-measure, and after some discussion, there does not seem to be support to continue to support the injection feature in its current form (or at all).

This issue is meant to track that concern and decide what to do. Among the options to consider are:

1. Normatively, continue to support context injection, and provide much more clear rules on how to support it across all cryptosuites.
2. Normatively, do not support context injection at all.
2. Normatively, do not prohibit context injection, but informatively note that its up to applications to do it after they have carefully considered the ramifications.

**Comment by jandrieu** [comment]
Can we get clarification on what "context injection" is?

**Comment by rflechtner** [comment]
> Can we get clarification on what "context injection" is?

Context injection is described in https://www.w3.org/TR/vc-data-integrity/#context-injection and refers to amending the `@context` property of the secured document with a context value that maps the terms contained in a proof. In practice, this means that the document is altered in the process of proof creation - or even of proof verification, the specification is not entirely clear here. My assumption is that this practice was initially introduced under the assumption that the `@context` property is not secured by the proof (see https://github.com/w3c/vc-data-integrity/issues/272 where this has been discussed) and thus can be manipulated without negative consequences, an assumption that however does not hold true when using cryptosuites that treat the document as regular JSON, e.g. those using the JCS transformation (such as [eddsa-jcs-2022](https://www.w3.org/TR/vc-di-eddsa/#eddsa-jcs-2022)).

We've tried addressing the issue by means of a workaround on JCS cryptosuites (https://github.com/w3c/vc-data-integrity/issues/225), but as I outlined in https://github.com/w3c/vc-data-integrity/issues/231 there are more accommodations that would need to be made to make sure that JCS-based proofs can reliably be verified. These accommodations, if they are even possible, need to be very carefully designed to make sure that not only reliably verification is guaranteed, but also an effective protection of the semantics expressed in `@context` values originally added by the issuer.

These problems uncover that the praxis of context injection leads to an unnecessarily tight coupling to JSON-LD. As I tried to describe in https://github.com/w3c/vc-data-integrity/issues/231#issuecomment-2211027752, abandoning context injection (the practical value of which is unclear) means that Linked Data becomes an additional semantic layer that _can_ be added to a document, but is not strictly required when using cryptosuites such as __eddsa-jcs-2022__.

**Comment by rflechtner** [comment]
I'll try to briefly summarise the arguments for and against abandoning context injection as a feature of this specification as far as I see them.

__Reasons to abandon context injection:__

- Context injection is the only feature of the DI specs as currently proposed that violates the reasonable assumption that the _document_ to be secured is not altered through the process of issuing a _proof_ for it; or, in other words, that information relating to cryptographic authentication should be distinguishable and expressed separately from the data it is meant to secure.
- Violating the above assumption is orthogonal to the concept of `proof sets` or `proof chains`, where multiple proofs secure the same document, often applied consecutively. If adding a _proof_ can change the _document_, then later proofs can effectively secure a different _document_ than the ones added earlier.
- The manipulation of the document's `@context` property as part of adding a proof always carries the possibility of edge cases where the semantics of the _document_ are changed simply by adding a _proof_, especially where JCS based cryptosuites are used.
- Abandoning context injection allows equal support for JCS- and LD/RDF-based cryptosuites and removes the need for a number of semi-functional workarounds we've had to apply to JCS-based cryptosuite specifications.
- This also means that the understanding and use of Linked Data semantics remains an option, but is no longer a hard requirement when using JCS-based cryptosuites. Not only does that extend the range of use cases for this specification, it will likely also make it much easier to win over Linked Data sceptics - not only to support and use this specification, but potentially even to transition to Linked Data semantics and features eventually, as there is a simple and straightforward 'upgrade path' available.

__Reasons against abandoning context injection:__

- Interoperability / Legacy Support: Given that it has been part of this specification for the longest time, and that similar proof methods have preceded it, context injection may performed as part of existing implementations. Designing the DI specifications on the assumption that the _document_ is not altered by adding proofs would mean that JCS-based proofs may be invalidated when legacy proof methods are added to a document later on - at least if we would also revert the workaround developed in https://github.com/w3c/vc-data-integrity/issues/225
- Context injection also serves the purpose of guaranteeing that the terms used in DI proofs can be reliably mapped to Linked Data and cannot be redefined. New measures would have to be introduced to enforce this; but solutions are readily available, for example in form of adding a `@context` property to the _proof_ object itself that provides the relevant mappings.

**Comment by msporny** [comment]
@rflechtner, some rough thoughts on potential normative language that we could use/add -- I haven't thought through each one in detail (some of them might be unworkable):

1. Injection MUST happen at the application layer, that's the only place it makes sense to do it.
2. The application layer MAY inject stuff BEFORE it adds ANY proof.
3. The application layer MUST/SHOULD NOT inject stuff before it it verifies a proof.
4. If the application layer injects a context, it MUST ensure that at least the data-integrity/v2 terms are a part of that injection (and maybe this doesn't apply to JCS)?

**Comment by rflechtner** [comment]
> some rough thoughts on potential normative language that we could use/add

1-3 make perfect sense to me.

> If the application layer injects a context, it MUST ensure that at least the data-integrity/v2 terms are a part of that injection 

This moves into the territory of what I referred to as 'alternative measures to enforce that proof terms can be mapped to Linked Data', and thus depends on the preferred solution. What you sketched out is one possible solution that does seem workable, but requires that the application logic that puts together the document to be secured considers the semantics of the proof that's added by a DI implementation.

I currently see roughly 3 possible types of solutions of which we may pick one:

- Require application logic to ensure correct mapping of proof terms by including appropriate contexts in the document `@context` property before any proofs are added; this may be mandated always, or could be required only if the application wants to use proofs that require RDF canonicalization.
- Let DI implementations enforce semantics of proof terms explicitly by including a `@context` property _on the proof object_ that includes the value `https://w3id.org/security/data-integrity/v2`. This could be made conditional of course, such that this `@context` value is only added if the document doesn't already map all the terms correctly.
- Make enforcing correct proof property mappings an implementation detail of RDFC-based proofs. The [_proof configuration_ algorithm](https://www.w3.org/TR/vc-di-eddsa/#proof-configuration-eddsa-rdfc-2022) of __eddsa-rdfc-2022__ and similar suites calls for "Setting proofConfig.@context to unsecuredDocument.@context" which could be amended to inject the value `https://w3id.org/security/data-integrity/v2`. This way of enforcing semantics is more implicit and does not strictly guarantee that the entire secured document (including proofs) can be properly understood by a JSON-LD processor.

Personally, I'm currently leaning towards preferring the second solution where the context mapping proof-related term is added to the proof object instead, simply for the separation of concerns that this affords: let the application logic worry about the semantics of the document (if it wants to deal with JSON-LD at all) and the DI implementation worry about enforcing semantics for its terms, without the two interacting. So far this seems workable to me, but it's also not 100% thought out.

In normative language, this could read as follows:
During proof creation, implementation MUST ensure the correct mapping of all proof terms by injecting the context value `https://w3id.org/security/data-integrity/v2` to the `@context` property of the _proof_ object. This injection MAY be omitted if the `@context` value of the _document_ already provides an equivalent mapping, e.g., when it uses the Verifiable Credential Data Model v2.0 context (`https://www.w3.org/ns/credentials/v2`).

**Comment by veikkoeeva** [comment]
> Let DI implementations enforce semantics of proof terms explicitly by including a @context property on the proof object that includes the value https://w3id.org/security/data-integrity/v2. This could be made conditional of course, such that this @context value is only added if the document doesn't already map all the terms correctly.

From an implementation perspective, considering the discussion this links to, this probably requires a bit more beefed up step-by-step algorithm with e.g. an example of a starting document showing trasformations and what pieces of it are signed and so on. I tell this also because I indeed implemented one version with .NET and it took some time to figure out that one version that seemed to work, even if I had "matters of confusion" (such as this `injection` discussed here). Some of the implementation issues I think I have in .NET were just recently like https://github.com/dotnetrdf/dotnetrdf/issues/615 (the official version having this was released just some weeks ago, see also https://github.com/dotnetrdf/dotnetrdf/issues/631) and so the simpler the better and easier to review and verify the individual given library used steps the better. .NET also does not have native JCS support, but that's easier to do and understand, I think.

**Comment by rflechtner** [comment]
> considering the discussion this links to, this probably requires a bit more beefed up step-by-step algorithm

This is certainly the end goal, but there's still quite a few different approaches we could choose from that would take us to different algorithms. The simplest approach, even if most verbose, is to just solve it on the level of the proof data model: Defining that a Data Integrity proof MUST have a `@context` property and that its value MUST be the string `https://w3id.org/security/data-integrity/v2` or an array of strings whose first item is `https://w3id.org/security/data-integrity/v2`.

In this case there's no conditional logic and no fancy algorithms needed; simply by requiring a context that defines all relevant terms to __always__ be present on the proof we've made sure that all its terms can be mapped. When I said that this _could_ be made conditional, I was referring to this trade-off between verbosity and algorithmic complexity.

As with the choice between the three general approaches that I outlined, the decision which path to take depends a lot on preference, so before I draft any concrete algorithm I would want to see some agreement among the Editors and stakeholders of the specification concerning _if_ and _how_ this change is desired.

**Comment by dlongley** [comment]
> The simplest approach, even if most verbose, is to just solve it on the level of the proof data model: Defining that a Data Integrity proof MUST have a `@context` property and that its value MUST be the string https://w3id.org/security/data-integrity/v2 or an array of strings whose first item is https://w3id.org/security/data-integrity/v2.

Unfortunately, this approach does not work for cryptosuites that protect the underlying information (claims) instead of a particular expression of the JSON. In other words, this approach would prevent being able to call standard JSON-LD compaction APIs to produce new expressions using different contexts -- without causing the proofs to become invalid. So this approach is a non-starter -- and is related to why context injection exists to begin with.

Another approach might be to say whether a cryptosuite requires context injection (the rdfc cryptosuites require this) or whether it will provide it itself (the jcs cryptosuites do this).

**Comment by TallTed** [comment]
All —

Please review your comments above, and edit them to codefence all instances of `@keyword`, including (especially!) those in content within your comment which was quoted from someone else.

All you need to do is wrap those `@keyword` instances in single backticks, like ``` `@keyword` ```.

Failure to do this loops various GitHub users (e.g., `@context`, `@vocab`) into our conversations without their permission nor request.

Please, let's be good gitizens.

**Comment by rflechtner** [comment]
> this approach would prevent being able to call standard JSON-LD compaction APIs to produce new expressions using different contexts -- without causing the proofs to become invalid

@dlongley Could you elaborate on where exactly things break? I'm struggling to see the need to transform a document's content using standard JSON-LD compaction/extension and similar as a blocker for this solution, and I suspect that drilling down on this may reveal mismatched assumption on where in the verify - validate - process pipeline JSON-LD transformation can actually happen.

For me, whose attention is focused on JCS-based proofs, it seems that any sort of transformation of a secured document – including those using JSON-LD mechanisms – should not be permissible prior to proof verification, as it will break any JCS proof that may be part of the proof set. If we want to set this specification up to really be agnostic to how cryptosuites secure the data contained within the document, I see no way around this.

If your application relies on producing new expressions of the data using JSON-LD apis, I'd expect that this must happen _after_ proof verification, at which point breaking proofs shouldn't be an issue any more; in fact, you may as well process only the _unsecured document_, i.e., drop the proof property before feeding the document to your JSON-LD processor.
 
> Another approach might be to say whether a cryptosuite requires context injection (the rdfc cryptosuites require this) or whether it will provide it itself (the jcs cryptosuites do this).

Limiting context injection to RDFC proofs unfortunately does not help, at least in its 'explicit' form (where the added context value is not just an implementation detail but persisted on the secured document) – it would still break any JCS proofs in the proof set. Morphing context injection to an 'implicit' procedure that does not effectively alter the document is equivalent to the third of three suggestions I made, for reference:

> - Make enforcing correct proof property mappings an implementation detail of RDFC-based proofs. The [_proof configuration_ algorithm](https://www.w3.org/TR/vc-di-eddsa/#proof-configuration-eddsa-rdfc-2022) of __eddsa-rdfc-2022__ and similar suites calls for "Setting `proofConfig.@context` to `unsecuredDocument.@context`" which could be amended to inject the value `https://w3id.org/security/data-integrity/v2`. This way of enforcing semantics is more implicit and does not strictly guarantee that the entire secured document (including proofs) can be properly understood by a JSON-LD processor.

**Comment by dlongley** [comment]
@rflechtner,

> If your application relies on producing new expressions of the data using JSON-LD apis, I'd expect that this must happen after proof verification, at which point breaking proofs shouldn't be an issue any more; in fact, you may as well process only the unsecured document, i.e., drop the proof property before feeding the document to your JSON-LD processor.

No, it needs to happen before proof verification, as this is a major feature of the RDFC cryptosuite approach. It enables preservation of the securing mechanism across multiple JSON (and even non-JSON) expressions of the same underlying fundamental VCDM claims (i.e., subject => globally unambiguous property => value). This allows (for example, but not limited to):

* translation of VCs, without having to go back to the issuer for permission, to other languages (English JSON keys => Japanese JSON keys),
* selective disclosure using international contexts but dropping regional-specific ones, privacy preserving changing of the context for specific-verifier consumption without notifying the issuer,
* translation of VCs to other syntaxes like CBOR-LD and back again, potentially using new contexts that enable far greater compression in binary form than the original context -- again, without having to send additional "original contextual information" and without permission / phone home or looping in the issuer.

When you run JSON-LD compaction, only a top-level `@context` for the new document is output. So requiring embedded `@context` in proofs is problematic for this approach. It's not an issue for JCS proofs because those only secure the particular JSON expression itself (not the underlying VCDM claims like RDFC, regardless of JSON expression) which necessarily prohibits changing that JSON expression. That works just fine for the use cases JCS proofs target where JSON expression does not need to change, but it does not cover the RDFC ones.

One more note: The `proof` property itself must always be defined. This implies the need for a top-level context that does this, regardless of JCS/RDFC cryptosuite.

**Comment by rflechtner** [comment]
> No, it needs to happen before proof verification, as this is a major feature of the RDFC cryptosuite approach.

@dlongley I worry that making this a core feature splits the specification in half. As you mentioned, anything that requires transforming the document at any point before verification (besides selective disclosure, maybe) won't work for JCS proofs.

Features like proof sets/proof chains and 'plug-in crypto' (i.e., cryptosuites) IMO suggest that handling credentials should be done in a manner that is agnostic to the type of proof that secures it, because except for closed-world scenarios you can't be certain which proofs are included and should ideally be able to verify multiple kinds.

In other words: in order to make sure that JCS proofs are equally supported, JSON-LD processing should be happening after verification. The best example is translation; I see zero reason why transforming a credential to use Japanese expressions can't happen after its proofs have been verified. Similarly, when it comes to transforming data to compact representations for size-efficient serialisation, I'm not sure it's a great idea to design compression around solutions that only work for some credentials but not all. As far as I can see a JSON-LD expansion and compaction cycle is never really guaranteed to result in the original data representation, so this just won't work for any proof mechanism that's based on a specific representation and not on the underlying linked data triples.

**Comment by dlongley** [comment]
I just noticed that we have a typo in the context injection section:

> If an @context property is not provided in a document that is being secured ***or verified*** ...  implementations MUST inject or add an @context property...

Context injection should NOT happen during verification, only when securing. During verification, a document should instead be rejected if it does not have a context that is known to define the data integrity terms. I wonder how much better things get if we fix this mistake? Does it resolve everything or do other troubles remain?

EDIT: I should also note that I think it's ok for different use cases to exist -- some that need document transformation and other cases that do not. We should not prohibit some use cases on account of others if we can avoid doing so. This spec  provides some basic primitives to developers to put together to solve their use cases and be innovative in ways we cannot predict; it is not meant to be prescriptive about use cases.

**Comment by rflechtner** [comment]
> Context injection should NOT happen during verification, only when securing. During verification, a document should instead be rejected if it does not have a context that is known to define the data integrity terms. I wonder how much better things get if we fix this mistake? Does it resolve everything or do other troubles remain?

@dlongley This is the issue that was discussed in https://github.com/w3c/vc-data-integrity/issues/231 and I did prepare a suggestion for a change and concretisation of these steps. I had not submitted these yet as a PR because I was not particularly happy with the result; while I managed to mitigate some issues, the solutions remain crude fixes that don't really address the underlying problem: that there is one step which the specification describes (adding a context that describes the proof object to the document) which directly contradicts the underlying assumptions of other aspects (that multiple proofs can secure the same document, in proof sets and proof chains). Clearly, if the _adding a proof_ routine tells you to change the document (at least conditionally) you'll end up in situations where proofs in a set do not secure _the same_ document, and that's where problems begin. 

More specifically, what I still wasn't able to solve:  
- The way we've worked around issues with context injection in JCS suites now means that it is possible to add additional context values to the end of the set of document contexts without verifiers being able to detect this. That makes me somewhat uncomfortable; it's not hard to foresee issues here where document terms that for example have not been properly defined in existing contexts are overloaded, post-issuance, with semantics the issuer did not intend.
- The mitigation strategy for the above as proposed in the specification – namely for verifiers to simply expect a fixed set of contexts in a specified order – also creates more questions than it answers. The order of context values seems hard to predict when it depends on the order in which proofs are added.
- This gets worse when considering that some contexts are to be treated as equivalent to the data integrity context (e.g., the VC v2.0 context). Issuers and verifiers would have to keep a list of 'equivalent contexts', and the possibility of these lists diverging would make it even harder to formulate reliable expectations about the set and order of context values.

To properly fix these issues, we'd have to go one step further, which takes us to a solution @msporny sketched out earlier in this thread (https://github.com/w3c/vc-data-integrity/issues/281#issuecomment-2211315321): require that the document context is finalised _before the first proof_ is added. This way JCS suites can build on the assumption that the document won't change, and lets them treat context values like any other data.
Thus, context injection would happen BEFORE securing; if a proof method detects that a document does not contain its required context, it should refuse proof creation instead of adding the context ad-hoc.

**Comment by msporny** [comment]
@rflechtner wrote:
> Thus, context injection would happen BEFORE securing; if a proof method detects that a document does not contain its required context, it should refuse proof creation instead of adding the context ad-hoc.

If we did this, would it completely address your concern?

Note this existing PR that removes context injection from any verification process and makes it only possible before adding a proof:

https://github.com/w3c/vc-data-integrity/pull/290/files#commit-suggestions

IOW, I think we are circling around this specific proposal you had above:
> Require application logic to ensure correct mapping of proof terms by including appropriate contexts in the document @context property before any proofs are added; this may be mandated always, or could be required only if the application wants to use proofs that require RDF canonicalization.

What other normative text would you like to see?

**Comment by rflechtner** [comment]
> If we did this, would it completely address your concern?

It would. Moving injection up to the application layer is not my preferred solution, but it's still a good compromise and solves the issues we're currently facing with JCS proofs.

> Note this existing PR that removes context injection from any verification process and makes it only possible before adding a proof

This does not yet clearly state that this injection is not allowed as soon as a document already contains one or more proofs. Ideally I would want to see this stated in the add proof / add proof set algorithms.

> IOW, I think we are circling around this specific proposal you had above

Exactly, and this is just a reformulation of what you had already proposed earlier:

> 1. Injection MUST happen at the application layer, that's the only place it makes sense to do it.
> 2. The application layer MAY inject stuff BEFORE it adds ANY proof.
> 3. The application layer MUST/SHOULD NOT inject stuff before it verifies a proof.
> 4. If the application layer injects a context, it MUST ensure that at least the data-integrity/v2 terms are a part of that injection 


> What other normative text would you like to see?

If implementations are required to reject any document that does not contain the data-integrity/v2 context _or equivalents_, then every implementation will have to have a list of _equivalent contexts_. Which contexts these are exactly is not specified. We could make interoperability more reliable by coming up with a process for aligning on and updating this list, for example a W3C-controlled registry.

**Comment by msporny** [comment]
> > If we did this, would it completely address your concern?
> 
> It would. Moving injection up to the application layer is not my preferred solution, but it's still a good compromise and solves the issues we're currently facing with JCS proofs.

Ok, got it. What would your preferred solution be? I read through the thread above and couldn't figure out your order of preferences (maybe I missed it?).

> > Note this existing PR that removes context injection from any verification process and makes it only possible before adding a proof
> 
> This does not yet clearly state that this injection is not allowed as soon as a document already contains one or more proofs. Ideally I would want to see this stated in the add proof / add proof set algorithms.

Ok, sound good.

> > IOW, I think we are circling around this specific proposal you had above
> 
> Exactly, and this is just a reformulation of what you had already proposed earlier:
> 
> > 1. Injection MUST happen at the application layer, that's the only place it makes sense to do it.
> > 2. The application layer MAY inject stuff BEFORE it adds ANY proof.
> > 3. The application layer MUST/SHOULD NOT inject stuff before it verifies a proof.
> > 4. If the application layer injects a context, it MUST ensure that at least the data-integrity/v2 terms are a part of that injection

Ok.

> > What other normative text would you like to see?
> 
> If implementations are required to reject any document that does not contain the data-integrity/v2 context _or equivalents_, then every implementation will have to have a list of _equivalent contexts_. Which contexts these are exactly is not specified. We could make interoperability more reliable by coming up with a process for aligning on and updating this list, for example a W3C-controlled registry.

I've got feedback from others in the social web community (not ActivityPub) that they would like the option of NOT having to inject a context at all. IOW, they want to use DI + JCS w/o having to inject a context (which is why it's optional to do so at the application layer). While this leads to undefined terms, or having to rely on text in specifications to understand the context, it is probably going to have to be a compromise we make to achieve broader adoption. In the ideal case, everyone injects their application-specific context (such as ActivityPub + DI), but we might have to hold our nose for those that don't want to do that, but still do want most of the benefits of using DI.

**Comment by rflechtner** [comment]
I realise now that there's actually 4 different proposals on the table, which I'll list in the order of my personal preference:

1. Do not enforce that proof terms are mapped; suggest that cryptosuites using RDFC mechanisms SHOULD inject the relevant contexts as part of proof/document transformation _such that it is not persisted_ (pure implementation detail of the transformation algorithm of those suites). This way signing documents with no context at all is allowed and possible with JCS suites (RDFC couldn't process the document though). If all terms on the document are mapped by the document `@context`, there is no hard requirement for the application layer to add the data-integrity/v2 context too, even for RDFC suites. 
2. Do not enforce that proof terms are mapped but suggest that the application layer SHOULD/MAY include/inject the data-integrity/v2 context in the unsigned document (before any proof is added). The difference to 1. is that it is the application layer's responsibility to add data-integrity/v2 or equivalent contexts to the unsecured document if RDFC proof are to be used to secure it.
3. Add the data-integrity/v2 context to the proof object instead. What I like about this option is that semantics of document and proof are expressed independently. It suffers from a shortcomings that Dave has outlined though: the `proof` property would remain undefined, and re-compacting or otherwise transforming the entire secured document would move the proof's `@context`.
4. Enforce that proof terms are mapped by saying that applications MUST add the data-integrity/v2 context (or equivalents) to the unsecured document, and that implementations MUST fail proof generation and verification if the (un)secured document does not contain a context that maps these terms. This is what I had in mind when saying that "moving injection up to the application layer is not my preferred solution", because it significantly raises the complexity for both applications and implementations.

In short, if your proposal was to make JSON-LD semantics an elective feature (for JCS cryptosuites) by adopting either solution 1 or 2, I'm very much on board.

**Comment by msporny** [comment]
> In short, if your proposal was to make JSON-LD semantics an elective feature (for JCS cryptosuites) by adopting either solution 1 or 2, I'm very much on board.

Yes, that was what I was proposing because it seemed to draw the strongest support and weakest objections among other people I'm speaking with in the social web community. I don't think it's either @dlongley or my preferred solution, but I think it maximizes the number of people that would be interested in using Data Integrity (we pull in the "Don't force me to use JSON-LD processors" people while signalling that "Even if you don't use `@context`, the semantics, whether you're explicit about it or not, are JSON-LD + Data Integrity." (Hope that makes sense, I haven't found a cleaner way to express it yet). 

I'm sure @dlongley might have some comments on the above, so I'll wait for him to add those before taking action in a PR.

**Comment by dlongley** [comment]
@msporny,

Some variant of option 2 seems ok to me. I think we could indicate that the DI properties can be used in a document that lacks `@context` (if desired), at the cost of being unprocessable as JSON-LD. I think that's fine for some use cases. In fact, if we just use this new language for the top-level `proof` property, it might be simplest. This is because we already nest `@context` in the JCS proofs today, allowing for JCS cryptosuite usage with JSON-LD documents if we don't change that. I think this approach might continue to make testing easier with the test suites we already have as well.

I think we should also say that extensions to the non-JSON-LD approach cannot define or use any new terms (that don't already appear in the DI specs) unless they also define a JSON-LD context so the parity continues. Given that we explicitly designed DI to be considerably extensible without the need for new terms anyway, I think this can work well (e.g., we have embraced encapsulation, like with multibase-CBOR-encoded `proofValue` values for cryptosuite-specific complexities).

**Comment by msporny** [comment]
PR #301 has been raised to address this issue. This issue will be closed once PR #301 has been merged.

**Comment by msporny** [comment]
PR #301 has been merged, closing.
